### PR TITLE
[System]: Fix issue in ServicePointScheduler with collection being modified in foreach.

### LINKED ISSUE #7355: "InvalidOperationException: Collection was modified; enumeration operation may not execute." during `ServicePointScheduler.RunSchedulerIteration()` when `HttpClient` is disposed after "NameResolutionFailure", "Access denied", or "Network is unreachable"
## Regression status: regression starting at cd09ddbc55c

`ServicePointScheduler` is a new class that was introduced in cd09ddbc55c. But maybe this new code itself isn't the root of the problem, and it is just revealing an earlier regression in a different part of Mono?

- Bad: [master@2c2fac29ebc](https://jenkins.mono-project.com/view/Per-commit%20packages/job/build-package-osx-mono/job/master/608/) (a build from today)
- Bad: [master@cd09ddbc55c](https://jenkins.mono-project.com/view/Per-commit%20packages/job/build-package-osx-mono/job/master/110/)
- Good: [master@7bf60c3ec95](https://jenkins.mono-project.com/view/Per-commit%20packages/job/build-package-osx-mono/job/master/106/)

## Steps to reproduce

1. Compile and run the following program:

    ```csharp
    using System;
    using System.Net.Http;
    using System.Threading.Tasks;
    using System.Threading;
    
    class Program {
    	public static void Main () {
    		Run ().Wait ();
    		Console.WriteLine ("done");
    		Thread.Sleep (500);
    	}
    
    	async static Task Run ()
    	{
    		try {
    			using (var client = new HttpClient ()) {
    				var response = await client.GetAsync ("http://255.255.255.255");
    			}
    		} catch (Exception ex) {
    		}
    	}
    }
    ```

Another way to hit the issue is to disable all of your network interfaces and adjust the `GetAsync()` call to use any non-loopback domain name or IP address.

It might be useful to mention that neither a "Connection refused" `WebException` nor a "Connection timed out" `WebException` seems to trigger the problem, but "NameResolutionFailure", "Access denied", and "Network is unreachable" `WebException` errors all do. It might also be interesting that the problem doesn't seem to happen unless the `HttpClient` is disposed.

## Current Behavior

An unhandled `InvalidOperationException` exception is thrown when the program is exiting:

```
done

Unhandled Exception:
Nested exception detected.
Original Exception: at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () [0x0000c] in /Users/builder/jenkins/workspace/build-package-osx-mono/master/external/bockbuild/builds/mono-x64/mcs/class/referencesource/mscorlib/system/runtime/exceptionservices/exceptionservicescommon.cs:155
at System.Runtime.CompilerServices.AsyncMethodBuilderCore/<>c.<ThrowAsync>b__6_1 (object) [0x00000] in /Users/builder/jenkins/workspace/build-package-osx-mono/master/external/bockbuild/builds/mono-x64/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/AsyncMethodBuilder.cs:1034
at System.Threading.QueueUserWorkItemCallback.WaitCallback_Context (object) [0x0000d] in /Users/builder/jenkins/workspace/build-package-osx-mono/master/external/bockbuild/builds/mono-x64/mcs/class/referencesource/mscorlib/system/threading/threadpool.cs:1308
at System.Threading.ExecutionContext.RunInternal (System.Threading.ExecutionContext,System.Threading.ContextCallback,object,bool) [0x00071] in /Users/builder/jenkins/workspace/build-package-osx-mono/master/external/bockbuild/builds/mono-x64/mcs/class/referencesource/mscorlib/system/threading/executioncontext.cs:961
at System.Threading.ExecutionContext.Run (System.Threading.ExecutionContext,System.Threading.ContextCallback,object,bool) [0x00000] in /Users/builder/jenkins/workspace/build-package-osx-mono/master/external/bockbuild/builds/mono-x64/mcs/class/referencesource/mscorlib/system/threading/executioncontext.cs:908
at System.Threading.QueueUserWorkItemCallback.System.Threading.IThreadPoolWorkItem.ExecuteWorkItem () [0x00021] in /Users/builder/jenkins/workspace/build-package-osx-mono/master/external/bockbuild/builds/mono-x64/mcs/class/referencesource/mscorlib/system/threading/threadpool.cs:1285
at System.Threading.ThreadPoolWorkQueue.Dispatch () [0x00074] in /Users/builder/jenkins/workspace/build-package-osx-mono/master/external/bockbuild/builds/mono-x64/mcs/class/referencesource/mscorlib/system/threading/threadpool.cs:858
at System.Threading._ThreadPoolWaitCallback.PerformWaitCallback () [0x00000] in /Users/builder/jenkins/workspace/build-package-osx-mono/master/external/bockbuild/builds/mono-x64/mcs/class/referencesource/mscorlib/system/threading/threadpool.cs:1213

Nested exception:at System.RuntimeTypeHandle.GetGenericTypeDefinition (System.RuntimeType) [0x00000] in /Users/builder/jenkins/workspace/build-package-osx-mono/master/external/bockbuild/builds/mono-x64/mcs/class/corlib/System/RuntimeTypeHandle.cs:157
at System.RuntimeType.GetGenericTypeDefinition () [0x00018] in /Users/builder/jenkins/workspace/build-package-osx-mono/master/external/bockbuild/builds/mono-x64/mcs/class/referencesource/mscorlib/system/rttype.cs:4320
at System.Diagnostics.StackTrace.GetFullNameForStackTrace (System.Text.StringBuilder,System.Reflection.MethodBase,bool,bool&) [0x00017] in /Users/builder/jenkins/workspace/build-package-osx-mono/master/external/bockbuild/builds/mono-x64/mcs/class/corlib/System.Diagnostics/StackTrace.cs:256
at System.Diagnostics.StackTrace.AddFrames (System.Text.StringBuilder) [0x00078] in /Users/builder/jenkins/workspace/build-package-osx-mono/master/external/bockbuild/builds/mono-x64/mcs/class/corlib/System.Diagnostics/StackTrace.cs:218
at System.Diagnostics.StackTrace.ToString () [0x0001c] in /Users/builder/jenkins/workspace/build-package-osx-mono/master/external/bockbuild/builds/mono-x64/mcs/class/corlib/System.Diagnostics/StackTrace.cs:323
at System.Diagnostics.StackTrace.ToString (System.Diagnostics.StackTrace/TraceFormat) [0x00000] in /Users/builder/jenkins/workspace/build-package-osx-mono/master/external/bockbuild/builds/mono-x64/mcs/class/corlib/System.Diagnostics/StackTrace.cs:340
at System.Environment.GetStackTrace (System.Exception,bool) [0x00014] in /Users/builder/jenkins/workspace/build-package-osx-mono/master/external/bockbuild/builds/mono-x64/mcs/class/corlib/System/Environment.cs:1036
at System.Exception.GetStackTrace (bool) [0x00038] in /Users/builder/jenkins/workspace/build-package-osx-mono/master/external/bockbuild/builds/mono-x64/mcs/class/referencesource/mscorlib/system/exception.cs:390
at System.Exception.ToString (bool,bool) [0x00081] in /Users/builder/jenkins/workspace/build-package-osx-mono/master/external/bockbuild/builds/mono-x64/mcs/class/referencesource/mscorlib/system/exception.cs:488
at System.Exception.ToString () [0x00000] in /Users/builder/jenkins/workspace/build-package-osx-mono/master/external/bockbuild/builds/mono-x64/mcs/class/referencesource/mscorlib/system/exception.cs:465


[ERROR] FATAL UNHANDLED EXCEPTION: System.InvalidOperationException: Collection was modified; enumeration operation may not execute.
  at System.Collections.Generic.Dictionary`2+Enumerator[TKey,TValue].MoveNext () [0x00013] in <d01c84b8049a4b51b4d4ef459d621744>:0 
  at System.Net.ServicePointScheduler.RunSchedulerIteration () [0x00046] in /Users/builder/jenkins/workspace/build-package-osx-mono/master/external/bockbuild/builds/mono-x64/mcs/class/System/System.Net/ServicePointScheduler.cs:248 
  at System.Net.ServicePointScheduler+<StartScheduler>d__32.MoveNext () [0x00228] in /Users/builder/jenkins/workspace/build-package-osx-mono/master/external/bockbuild/builds/mono-x64/mcs/class/System/System.Net/ServicePointScheduler.cs:174 
--- End of stack trace from previous location where exception was thrown ---
  at System.Runtime.CompilerServices.AsyncMethodBuilderCore+<>c.<ThrowAsync>b__6_1 (System.Object state) [0x00000] in /Users/builder/jenkins/workspace/build-package-osx-mono/master/external/bockbuild/builds/mono-x64/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/AsyncMethodBuilder.cs:1034 
  at System.Threading.QueueUserWorkItemCallback.WaitCallback_Context (System.Object state) [0x0000d] in /Users/builder/jenkins/workspace/build-package-osx-mono/master/external/bockbuild/builds/mono-x64/mcs/class/referencesource/mscorlib/system/threading/threadpool.cs:1308 
  at System.Threading.ExecutionContext.RunInternal (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state, System.Boolean preserveSyncCtx) [0x00071] in /Users/builder/jenkins/workspace/build-package-osx-mono/master/external/bockbuild/builds/mono-x64/mcs/class/referencesource/mscorlib/system/threading/executioncontext.cs:961 
  at System.Threading.ExecutionContext.Run (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state, System.Boolean preserveSyncCtx) [0x00000] in /Users/builder/jenkins/workspace/build-package-osx-mono/master/external/bockbuild/builds/mono-x64/mcs/class/referencesource/mscorlib/system/threading/executioncontext.cs:908 
  at System.Threading.QueueUserWorkItemCallback.System.Threading.IThreadPoolWorkItem.ExecuteWorkItem () [0x00021] in /Users/builder/jenkins/workspace/build-package-osx-mono/master/external/bockbuild/builds/mono-x64/mcs/class/referencesource/mscorlib/system/threading/threadpool.cs:1285 
  at System.Threading.ThreadPoolWorkQueue.Dispatch () [0x00074] in /Users/builder/jenkins/workspace/build-package-osx-mono/master/external/bockbuild/builds/mono-x64/mcs/class/referencesource/mscorlib/system/threading/threadpool.cs:858 
  at System.Threading._ThreadPoolWaitCallback.PerformWaitCallback () [0x00000] in /Users/builder/jenkins/workspace/build-package-osx-mono/master/external/bockbuild/builds/mono-x64/mcs/class/referencesource/mscorlib/system/threading/threadpool.cs:1213 
```

## Expected Behavior

With mono/master@7bf60c3ec95 or .NET Core 2.1.4, the program exits with no exceptions.

## Additional testing environment info (brief)

macOS 10.12.6
US English locale

### On which platforms did you notice this

\[x\] macOS
\[ \] Linux
\[ \] Windows

### PR DESCRIPTION
**Pull request by baulig**
Fixes #7355

### CODE REVIEW

File: mcs/class/System/System.Net/ServicePointScheduler.cs [hunk 1]
```diff
@@ -245,9 +245,14 @@ void RunSchedulerIteration ()
 				Debug ($"ITERATION #1: {repeat} {groups != null}");
 
 				if (groups != null) {
-					foreach (var group in groups) {
-						Debug ($"ITERATION #2: group={group.Value.ID}");
-						repeat |= SchedulerIteration (group.Value);
+					var keys = new string[groups.Count];
```
Comment (marek-safar) [inline mcs/class/System/System.Net/ServicePointScheduler.cs]: Doesn't this just hide the race bellow where you do `groups.Remove (groupName);` without a lock?

**Comment by marek-safar** [comment]
Please also add the test from the bug report to the PR

**Comment by baulig** [comment]
This PR should land _after_ #7358 because that's introducing a critical behavior change that will make this work reliably.

File: mcs/class/System/System.Net/ServicePointScheduler.cs (hunk 1 above)
Comment (baulig) [reply mcs/class/System/System.Net/ServicePointScheduler.cs]: There is no "race condition" here in the traditional sense and no locks or concurrent dictionaries are going to help you here.

What's happening here is that `SchedulerIteration()` might actually remove the current group from the list when it fails to establish a connection.

Prior to the commit that I just pushed on top of #7358, there as also a separate issue where arbitrary user code might get executed from within `SchedulerIteration()` (on our thread, while holding the lock) as part of a `TaskCompletionSource`s continuation.

This has now been fixed by replacing all uses of `TaskCompletionSource` with `WebCompletionSource`, which uses `TaskCreationOptions.RunContinuationsAsynchronously`.

**Comment by baulig** [comment]
Closing; as this should really land together with the changes from #7358, I cherry-picked this onto that PR and also added the test there.

**Pull request closed without merging**